inline values as decorators when debugging

nojvek opened the pull request:
1. [x] Write Tests.
2. [x] Figure out how to get words from lines. 
3. [x] Figure out how to stop flashing on stepping because onDidContinue passes in null stack frame everytime
4. [x] Figure out how to registerDecorator from public API. 
5. [x] setDecorators afterText doesn't work in python on lines with trailing ':'
6. [x] Merge as single commit
7. [x] Add screenshot

![image](https://cloud.githubusercontent.com/assets/1018196/20868173/fc17554a-ba09-11e6-9bc9-5acd16af8e99.png)

nojvek commented:
@isidorn Please take a look at the first iteration.

isidorn commented:
I will do a code review and provide feedback in the code a bit later today, let me first try to answer your questions
1. Awesome - might be a bit tricky though
2. Just checked and yes there is no nice API to get you all words on a line. As you figured out already you need to use a regex, however you should use a word regex for the particular language you are in. You should get the regex using this [method](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/modes/languageConfigurationRegistry.ts#L221). Here is a [sample usage](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/services/editorWorkerServiceImpl.ts#L349) 
3. Instead of using `editor.removeDecorations` and `editor.addDecorations` you need to use `editor.deltaDecorations` and probably keep track on your side of all the decorations from a previous state. DebugEditorModelManager is doing something similar now with breakpoints [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/browser/debugEditorModelManager.ts#L129) 
4. Why do you want to register a decorator from a public api? This feature will live in the vsccode core repo
5. No clue about that, we can investigate if it is a bug in the editor

isidorn commented:
2. However if your name regex is more precise than the language word regex then we should probably use that

4. Here is a [public api](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L979) to set decorations in case you need it for some other use case

isidorn commented:
@nojvek just to let you know that next week we have an endgame week - which means we take no new features. So it would be cool if we would try to land this feature together sometime this week. I have time on Thursday european time so I could try to wrap it up if something is still needed to be done then. Hope that works for you.

nojvek commented:
Hi Isidor,

That's the plan. I'm going to send out a new iteration in a bit. Got some
bugs fixed. Fortunately I have Thursday and Friday off so can work on those
days full time.


On Tue, Nov 29, 2016 at 12:50 AM, Isidor Nikolic <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> just to let you know that next week
> we have an endgame week - which means we take no new features. So it would
> be cool if we would try to land this feature together sometime this week. I
> have time on Thursday european time so I could try to wrap it up if
> something is still needed to be done then. Hope that works for you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263511368>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVNavCqUQkZbj2KFJJmOs4GxZ8IpDks5rC-dvgaJpZM4K9bza>
> .
>

nojvek commented:
Hi Isidor,

I'm in need of using proper ES6 Map and Set. Can I change the typescript
tsconfig to include the es2015.collection lib ? The version of electron
vscode uses seems to have Set and Map defined so it should be fine right?



On Tue, Nov 29, 2016 at 1:08 AM, Noj Vek <nojvek@gmail.com> wrote:

> Hi Isidor,
>
> That's the plan. I'm going to send out a new iteration in a bit. Got some
> bugs fixed. Fortunately I have Thursday and Friday off so can work on those
> days full time.
>
>
> On Tue, Nov 29, 2016 at 12:50 AM, Isidor Nikolic <notifications@github.com
> > wrote:
>
>> @nojvek <https://github.com/nojvek> just to let you know that next week
>> we have an endgame week - which means we take no new features. So it would
>> be cool if we would try to land this feature together sometime this week. I
>> have time on Thursday european time so I could try to wrap it up if
>> something is still needed to be done then. Hope that works for you.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263511368>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA-JVNavCqUQkZbj2KFJJmOs4GxZ8IpDks5rC-dvgaJpZM4K9bza>
>> .
>>
>
>

isidorn commented:
Not sure if we can use it in the core
@jrieken is the right person to answer this

jrieken commented:
re https://github.com/Microsoft/vscode/pull/16129#issuecomment-263522996 - Where do you need those? In an extension and the main-body of source code?

isidorn commented:
@jrieken main-body of source code

nojvek commented:
For time being I've added a map-set.d.ts in the typings folder.

The reason why I'm having hard time with using {} as map is that, I can't
quite set hasOwnProtype or prototype as keys, since they are already
defined properties. A scope variable could be this special object keys.

I've tested with python and javascript. Inline values work well so far.

Need to iron out a couple of bugs and clean up the registerDecorator call.

Its 3am here, I'm heading to bed.

Regards.

On Tue, Nov 29, 2016 at 2:57 AM, Isidor Nikolic <notifications@github.com>
wrote:

> @jrieken <https://github.com/jrieken> main-body of source code
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263539940>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVJBgevvbkPN02UuoMHRmao2JwqYUks5rDAUjgaJpZM4K9bza>
> .
>

jrieken commented:
> For time being I've added a map-set.d.ts in the typings folder.

That is not allowed :no_good_man: We have some technical debt there but in essence it would make ES6 Maps/Sets also available in the editor/base layer which we program against older runtime targets that don't support ES6. You might have a local version of those definitions, we have done that in a few [places](https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/workbench/api/node/mainThreadHeapService.ts#L16) but it is not nice. Also it seems you don't need that.

> The reason why I'm having hard time with using {} as map is that, I can't
quite set hasOwnProtype or prototype as keys, since they are already
defined properties. A scope variable could be this special object keys.

The fix for that is to *not* use `{}` but `Object.create(null)` or the utils defined in [`collections.ts`](https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/base/common/collections.ts#L23)

nojvek commented:
I see. I'll undo the change and use collections helper. Thanks for the tip.

On Tuesday, November 29, 2016, Johannes Rieken <notifications@github.com>
wrote:

> For time being I've added a map-set.d.ts in the typings folder.
>
> That is not allowed 🙅‍♂️ We have some technical debt there but in
> essence it would make ES6 Maps/Sets also available in the editor/base layer
> which we program against older runtime targets that don't support ES6. You
> might have a local version of those definitions, we have done that in a few
> places
> <https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/workbench/api/node/mainThreadHeapService.ts#L16>
> but it is not nice. Also it seems you don't need that.
>
> The reason why I'm having hard time with using {} as map is that, I can't
> quite set hasOwnProtype or prototype as keys, since they are already
> defined properties. A scope variable could be this special object keys.
>
> The fix for that is to *not* use {} but Object.create(null) or the utils
> defined in collections.ts
> <https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/base/common/collections.ts#L23>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263548319>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVBJqjfE4Bm6Hpwr4gVeV9IWdpuG4ks5rDA7MgaJpZM4K9bza>
> .
>

nojvek commented:
Another question.

Having the inline values besides comments kind of looks weird. We match
against all words. Is it okay for me to getLineTokens for each line and
ignore the comment tokens?



On Tuesday, November 29, 2016, Noj Vek <nojvek@gmail.com> wrote:

> I see. I'll undo the change and use collections helper. Thanks for the
> tip.
>
> On Tuesday, November 29, 2016, Johannes Rieken <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> For time being I've added a map-set.d.ts in the typings folder.
>>
>> That is not allowed 🙅‍♂️ We have some technical debt there but in
>> essence it would make ES6 Maps/Sets also available in the editor/base layer
>> which we program against older runtime targets that don't support ES6. You
>> might have a local version of those definitions, we have done that in a few
>> places
>> <https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/workbench/api/node/mainThreadHeapService.ts#L16>
>> but it is not nice. Also it seems you don't need that.
>>
>> The reason why I'm having hard time with using {} as map is that, I can't
>> quite set hasOwnProtype or prototype as keys, since they are already
>> defined properties. A scope variable could be this special object keys.
>>
>> The fix for that is to *not* use {} but Object.create(null) or the utils
>> defined in collections.ts
>> <https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/base/common/collections.ts#L23>
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263548319>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA-JVBJqjfE4Bm6Hpwr4gVeV9IWdpuG4ks5rDA7MgaJpZM4K9bza>
>> .
>>
>

isidorn commented:
That sounds like a decent workaround for now. So +1
I believe tokenization is already dealing with words so it would be cool if you could just get tokens and not do any additional word magic on a line (which you are currently doing).

nojvek commented:
Thanks.

Since getWordsRanges will mostly be creating a map from line tokens, makes
sense to move the  function to TextEditorModel and expose as a public right?

I wonder how I can keep track of when the editor model has been changed so
I can invalidate the cached wordsRangesMap



On Tuesday, November 29, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> That sounds like a decent workaround for now. So +1
> I believe tokenization is already dealing with words so it would be cool
> if you could just get tokens and not do any additional word magic on a line
> (which you are currently doing).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263635668>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVGgu3fpsvG3wieJdo9zTakYtjemfks5rDF5hgaJpZM4K9bza>
> .
>

isidorn commented:
@nojvek I would not move that to the TextEditorModel before I check with @alexandrudima. Will talk to him tomorrow in the office.

You can add a listener to the editor on changeModel. [Example](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts#L152)

isidorn commented:
Ok I have discussed this with @alexandrudima and there are concerns
* The current implementation of putting decorations after text is suboptimal and can hurt performance
* Going through each word on a line can be very performant for very long lines
* What do we do for people that have word wrapping enabled. Via that setting they have specifically enabled lines not to overflow over some border. 

Due to the above, this feature will be under a setting which by default is false. Once we have no more performance concerns we can lift this flag. @nojvek you do not have to worry about introducing this flag, I can just put it on top once we polish this pr.

Also we need to put a limit on the line size and just do not compute anything for very long lines. Since currently we are matching the words on a line to a set of variables, we could also do it the other way around - match variables across a line. However this strategey does not work best for short lines which are the most common in practice. Thus I believe for starters we should just ignore long line.

As for the word wrap issue - that is something we need to think about. Might not require immediate action.

In a perfect world the debug extension would have a better ast understanding of the file - this knowledge would come from a language server but we currently do not have this. @weinand might know better if we plan to investigate into this any time in the future

Due to all of the above we do not have to rush this feature in for November. But let's see how it goes.

nojvek commented:
1. Currently we only process max 100 variable scopes for perf
2. Every scope value is trimmed to max 50 chrs and ellipses added if too
long.
3. Maximum length of inline decorator will be no more than 150. All this
are constants that can easily be changed.
4. Adding this under feature switch makes sense. We can get good feedback
from beta users.
5. Not sure how to handle word wrapping but I believe that will be more of
a decorator implementation fix.
6. This might seem like a small feature but it makes a significant
difference in debugging experience. I'll do my best to get something
performant and usable.

On Wednesday, November 30, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> Ok I have discussed this with @alexandrudima
> <https://github.com/alexandrudima> and there are concerns
>
>    - The current implementation of putting decorations after text is
>    suboptimal and can hurt performance
>    - Going through each word on a line can be very performant for very
>    long lines
>    - What do we do for people that have word wrapping enabled. Via that
>    setting they have specifically enabled lines not to overflow over some
>    border.
>
> Due to the above, this feature will be under a setting which by default is
> false. Once we have no more performance concerns we can lift this flag.
> @nojvek <https://github.com/nojvek> you do not have to worry about
> introducing this flag, I can just put it on top once we polish this pr.
>
> Also we need to put a limit on the line size and just do not compute
> anything for very long lines. Since currently we are matching the words on
> a line to a set of variables, we could also do it the other way around -
> match variables across a line. However this strategey does not work best
> for short lines which are the most common in practice. Thus I believe for
> starters we should just ignore long line.
>
> As for the word wrap issue - that is something we need to think about.
> Might not require immediate action.
>
> In a perfect world the debug extension would have a better ast
> understanding of the file - this knowledge would come from a language
> server but we currently do not have this. @weinand
> <https://github.com/weinand> might know better if we plan to investigate
> into this any time in the future
>
> Due to all of the above we do not have to rush this feature in for
> November. But let's see how it goes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-263867455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVI5xW4VReDUiJb-0wcDlM-uSbHu4ks5rDXNigaJpZM4K9bza>
> .
>

isidorn commented:
All those constraints make sense but we should introduce another one - the length of the editor line that is being processed. Because if the lenght of that line is quite large you will still go through each word and try to find it in the variables view, so some constraint needs to be added there.
We can handle word wrapping later, not for version 1.

fyi I am on vacation on Friday but I plan to code review, merge this feature and try to wrap it up on Monday.

nojvek commented:
Sounds good. I'll skip tokenization if line length is greater than 1000
chars.

@alexandrudima <https://github.com/alexandrudima>, on getLineTokens there
is a inaccurateTokens parameter. What does that do?

Also what is the perf impact of getLineTokens for every line in a file?

On Thursday, December 1, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> All those constraints make sense but we should introduce another one - the
> length of the editor line that is being processed. Because if the lenght of
> that line is quite large you will still go through each word and try to
> find it in the variables view, so some constraint needs to be added there.
> We can handle word wrapping later, not for version 1.
>
> fyi I am on vacation on Friday but I plan to code review, merge this
> feature and try to wrap it up on Monday.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-264118939>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVBCV2x9u3sMmcVhcHL4gq-pDCk7Yks5rDo6QgaJpZM4K9bza>
> .
>

nojvek commented:
I've moved most of helper functions to debugInlineDecorators.ts where the core of logic really happens. Added 100% line, branch and function coverage for that file. 

I've tested on a bunch of different project types and it looks good so far. I've added a screenshot. 

Let me know if you need more changes. Its a nice sunny sunday, so I'll get back to this on Monday evening. Can't wait for this to get into nightly releases.

isidorn commented:
@nojvek great work! And nice that you tested it in different projects!

I have added comments in the commit directly.

Let's move this feature out to the next release because there is no need for the rush and I would love the insiders to self host on this feature for a couple of weeks. This will also give us time to nicely adress the issues I brought up in the review.

nojvek commented:
Hi Isidor, 

No problems. I am not seeing any comments in the review though. 

When do you think it will make insiders?

Regards.

isidorn commented:
@nojvek ok that explains why you did not react on my previous comment from a couple of days ago.
Here is the direct link to my comments, you should be able to nicely see them
https://github.com/Microsoft/vscode/pull/16129/commits/d36a042fe25635191ab5f0c0bed987f060fda2cf

It should make insiders in about 8, 9 days when we wrap up our endgame.

isidorn commented:
@nojvek my appologies - I did not click on submit review both times! I was used to using the old github UX which did not require an explict click on submit

nojvek commented:
😳 still not able to see them.

May be paste the comments in conversation window.

Thanks.

On Monday, December 5, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> ok that explains why you did not
> react on my previous comment from a couple of days ago.
> Here is the direct link to my comments, you should be able to nicely see
> them
> d36a042
> <https://github.com/Microsoft/vscode/pull/16129/commits/d36a042fe25635191ab5f0c0bed987f060fda2cf>
>
> It should make insiders in about 8, 9 days when we wrap up our endgame.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-264903633>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVMLuN1JBGcxcweLy2U_1P_WPyIyoks5rFD2MgaJpZM4K9bza>
> .
>

isidorn commented:
@nojvek you should be able to see them now, I made a mistake of not publishing them. Sorry for the misunderstanding!

nojvek commented on src/vs/editor/browser/services/codeEditorServiceImpl.ts:
```diff
@@ -310,7 +310,7 @@ class DecorationRenderHelper {
 				cssTextArr.push(strings.format(this._CSS_MAP.contentIconPath, opts.contentIconPath.toString(true).replace(/'/g, '%27')));
```
This was actually a VSCode bug that I discovered. You couldn't have after/before text decorators with a `'` in them. `_CSS_MAP.contentText` is defined as `contentText: 'content:\'{0}\';',` . This means the escaping of `"` actually did nothing. It should have been escaping single quotes `'`.  I'm fairly certain that this doesn't have side effects because its only escaping the contentText property and that's the only usage of it. 

Without this fix,python string values won't show up. Since python strings show up as `'cat bat mat'` in the watches view.

isidorn replied:
Ok makes sense. But the best thing would be to provide a seperate PR that only fixes this issue.
That PR I will then assign to the editor people. As the description just copy this comment.

nojvek replied:
Migrated to separate PR. please see https://github.com/Microsoft/vscode/pull/18361/files

nojvek commented on src/vs/editor/common/commonCodeEditor.ts:
```diff
@@ -622,6 +622,10 @@ export abstract class CommonCodeEditor extends EventEmitter implements editorCom
 		return this.model.deltaDecorations(oldDecorations, newDecorations, this.id);
 	}
 
```
setDecorators actually uses deltaDecorators. Inline decorators need to use after.contentText which ends up translating to a css:after block. For each of the unique decorators we need to create the selectors. By not using setDecorators we'll have to do that manually which will duplicate the logic in two places. 

setDecorators ends up calling `provider = new DecorationSubTypeOptionsProvider(this._styleSheet, key, parentTypeKey, options);` eventually.

setDecorators does the correct thing and provides the proper abstraction. Its just that before calling setDecorators, the decoratorTypeKey to be registered. That's why I added the registerDecoratorType api. 

I've spent hours trying to figure out if there was a nicer way to register the decoratorType. This seemed like the least obtrusive way since `this._registerDecorationType` was already being used.

The other workaround I can think is for setDecorators to auto-register the key if it doesn't exist.

isidorn replied:
Thanks for the explanation!
@aeschli how are you registering your folding end of line decorators? Can you provide a code pointer? Does it make sense to expose register editor decoration API as @nojvek is proposing here.

nojvek commented on src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts:
```diff
@@ -192,6 +200,42 @@ export class DebugEditorContribution implements IDebugEditorContribution {
+		// This means we'll get a null stackFrame followed quickly by a valid stackFrame.
+		// Removing all decorators and adding them again causes a noticeable UI flash due to relayout and paint.
+		// We want to only remove inline decorations if a null stackFrame isn't followed by a valid stackFrame in a short interval.
+		clearTimeout(this.removeDecorationsTimeoutId);
```
Explained the issue about deltaDecorators on other thread.  We need the timeout anyway since setDecorators gives us a nice diff for free. We just need to debounce the null between the continues, otherwise we end up removing all decorators and creating them again rather than only changing the subset that actually are different.

isidorn replied:
I understand that but this is still not a nice solution. Though it is fine for the first version and once we merge this in next week I can try to polish this if possible

nojvek commented on src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts:
```diff
@@ -0,0 +1,212 @@
+/*---------------------------------------------------------------------------------------------
```
👍

nojvek commented on src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts:
```diff
@@ -0,0 +1,212 @@
+}
+
+// Simple tokenizer that separates comments from words
+function mockLineTokens(lineContent: string): LineTokens {
```
I tried that. The issue is the tokens are whole line tokens since the `tokenizerSupport` is `null`  I tried using `model.setMode('javascript')` but that needs the javascript tokenizersupport to be registered and I couldn't fit all of the pieces to test that comments are ignored. So I wrote a simple stub that mocks `editor.getLineTokens()`.

Happy to change this if there is a way to plug in the javascript tokenizer.

isidorn replied:
@aeschli do we have some mock tokenizer which we could reuse here?

nojvek commented on src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts:
```diff
@@ -0,0 +1,167 @@
+export const ELLIPSES = '…';
+// LanguageConfigurationRegistry.getWordDefinition() return regexes that allow spaces and punctuation characters for languages like python
+// Using that approach is not viable so we are using a simple regex to look for word tokens.
+export const WORD_REGEXP = /[\$\_A-Za-z][\$\_A-Za-z0-9]*/g;
```
\b doesn't work since it doesn't consider `$` and `_`  as `word` character. Powershell, php would break seriously. Javascript allows $ and _ in variable names so they would have quirks as well.

isidorn replied:
Ok makes sense. Though I hate introducing new regexes. Once we merge this in I can investigate if we can reuse some other word logic we have

nojvek commented on src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts:
```diff
@@ -0,0 +1,167 @@
+
+export function getDecoratorFromNames(lineNumber: number, names: string[], nameValueMap: IStringDictionary<string>, linesContent: string[]): IDecorationOptions {
+	const margin = '10px';
+	const backgroundColor = 'rgba(255,200,0,0.2)';
```
The generated parent classname with registerDecorationType looks like `ced-inlineDecorator-4` I could add the parent `:after` selector but it might look weird. 

I think once the decorator apis use an underlying implementation using spans rather than css :after selectors, we can nicely add the chrome like syntax highlighting for the inline values and move this to css. wdyt?

isidorn replied:
We would like to change the underlying implementation since it is already causing performance issues for folding. Though I will have to sync with @alexandrudima on this

isidorn commented:
@nojvek thanks for your replies - we are in the middle of the endgame so I will reply to your comments in a couple of days once the bugfixing fuss is over. I also need to consult with @alexandrudima for the editor side of things concerning some of the questions you have raised.

nojvek commented:
Sounds good.

On Tuesday, December 6, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> thanks for your replies - we are in
> the middle of the endgame so I will reply to your comments in a couple of
> days once the bugfixing fuss is over. I also need to consult with
> @alexandrudima <https://github.com/alexandrudima> for the editor side of
> things concerning some of the questions you have raised.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-265129525>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVHdcEN7I79xXexkpYxlJHkGb0e0yks5rFUnqgaJpZM4K9bza>
> .
>

nojvek commented on src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts:
```diff
@@ -183,19 +193,104 @@ export class DebugEditorContribution implements IDebugEditorContribution {
+		// TODO: onFocusStackFrame isn't called for the first time when editor url changes due to source mapping
+		//console.log("onFocusStackFrame", this.editor.getModel().uri.toString(), this.editor.getModel().uri.toString(), stackFrame);
+
+		if (!stackFrame || stackFrame && stackFrame.source.uri.toString() !== this.editor.getModel().uri.toString()) {
```
It seems you commented on an older review. Does not apply to current the iteration .

isidorn commented:
@nojvek feel free to ignore my comments from couple of days ago which I accidently only released yesterday

isidorn commented:
@nojvek added some comments. Seems like it is in a good state, I would lke to merge it in some time next week and then I can look into improving some of the issues we raised in the discussion (end of line decorations, using timeouts, creating a tokenizer for tests). Most of the team is still busy with the endgame so I will sync with them next week for input on improving implentation of decorations

nojvek commented:
Thanks for the update @isidor

On Friday, December 9, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> added some comments. Seems like it is
> in a good state, I would lke to merge it in some time next week and then I
> can look into improving some of the issues we raised in the discussion (end
> of line decorations, using timeouts, creating a tokenizer for tests). Most
> of the team is still busy with the endgame so I will sync with them next
> week for input on improving implentation of decorations
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-266020696>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVPLun1T_0uXSP1WeT0HqQB4tS7U5ks5rGV-WgaJpZM4K9bza>
> .
>

nojvek commented:
@aeschli, In this instance managing the css on my own will probably be
almost the same as what setDecorators will do. The inline value decorators
are all after.contentText decorators which mean each of them will have a
corresponding ::after {..} css snippet.

That is why I wanted to add a registerDecoratorType api to the editor.

On Fri, Dec 9, 2016 at 8:48 AM, Martin Aeschlimann <notifications@github.com
> wrote:

> *@aeschli* commented on this pull request.
> ------------------------------
>
> In src/vs/editor/common/commonCodeEditor.ts
> <https://github.com/Microsoft/vscode/pull/16129>:
>
> > @@ -622,6 +622,10 @@ export abstract class CommonCodeEditor extends EventEmitter implements editorCom
>  		return this.model.deltaDecorations(oldDecorations, newDecorations, this.id);
>  	}
>
>
> https://github.com/Microsoft/vscode/blob/master/src/vs/
> editor/contrib/folding/common/foldingModel.ts#L71 is where folding
> defines its decorations. Note that it manages the CSS itself. It just
> passes in the name of the CSS in inlineClassName. You can also do
> before/after images & text the same way using beforeContentClassName/
> afterContentClassName
> If you do that (manage your own CSS) all you need to use deltaDecorators.
>
> setDecoration got added for our extension host API, where the CSS is
> managed by us (done in the CodeEditorService). If you want to use that API,
> register your decoration type on the ICodeEditorService. No new API should
> be required.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVGIb-yKbR95etGGRYAwJ7RtNoAa3ks5rGYZkgaJpZM4K9bza>
> .
>

isidorn commented:
Hi @nojvek I was looking into mergin this in, the following would still be needed in order for me to do the merge

- [ ] As @aeschli mentioned [here](https://github.com/Microsoft/vscode/pull/16129#discussion_r91751609) no new editor api needs to be added. So please use existing API
- [ ] Changes in the `codeEditorServiceImpl.ts` need to be moved to a seperate PR
- [ ] merge your branch with latest from master and resolve conflicts

Other things I can polish up once we merge it in.

Thanks a lot!

nojvek commented:
Sounds good. I'll get on it.

@aeschli, what did we decide regarding registerDecorator API ? It should
auto register key when calling setDecorators?

Regards.

On Wednesday, December 28, 2016, Isidor Nikolic <notifications@github.com>
wrote:

> Hi @nojvek <https://github.com/nojvek> I was looking into mergin this in,
> the following would still be needed in order for me to do the merge
>
>    - As @aeschli <https://github.com/aeschli> mentioned here
>    <https://github.com/Microsoft/vscode/pull/16129#discussion_r91751609>
>    no new editor api needs to be added. So please use existing API
>    - Changes in the codeEditorServiceImpl.ts need to be moved to a
>    seperate PR
>    - merge your branch with latest from master and resolve conflicts
>
> Other things I can polish up once we merge it in.
>
> Thanks a lot!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-269479138>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVDuljVAsci0kX33wCg0RSfGFNjcZks5rMmhggaJpZM4K9bza>
> .
>

aeschli commented:
My suggestion is that you register your decoration type using ICodeEditorService.registerDecorationType

isidorn commented:
ping @nojvek just to check if there are some updates here? Just fyi If you are busy I can also work on this.

nojvek commented:
@aeschli , codeEditorService is a private variable of codeEditor. It is not exposed. This is exactly what I am trying to do.  ICodeEditor.registerDecorationType ends up calling ICodeEditorService.registerDecorationType

Updated PR, as you recommend I am using ICodeEditorService.registerDecorationType `(<any>this.editor)._codeEditorService.registerDecorationType`. I get a feeling that you are very much against touching the code editor api.

I feel its worse solution than what we had before but gets the job done until vscode exposes a nicer api. I'm really not sure what you want me to do here. 😞

aeschli commented:
@nojvek To get hold of the codeEditorService, use service injection in the class you are:
Add
`@ICodeEditorService codeEditorService: ICodeEditorService`
to the constructor of the DebugEditorContribution

nojvek commented:
Whoa! Code injection dark magic.

Will give it a shot.

Is there any other major changes you need? I'd love to get this on this
week.

Will be on vacation until February.

Thanks.

On Tue, Jan 10, 2017 at 4:52 PM Martin Aeschlimann <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> To get hold of the codeEditorService,
> use service injection in the class you are:
>
>
> Add
>
>
> @ICodeEditorService codeEditorService: ICodeEditorService
>
>
> to the constructor of the DebugEditorContribution
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-271745563>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVMd50pQQJzg0e9YHcZEkksubCtNcks5rRCfegaJpZM4K9bza>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

isidorn commented:
@nojvek just checked the PR and no other major changes are needed - except the code injection Martin mentioned. Once that is in I will merge this in.
Some polish work will be needed on top (adding setting, some refactoring, trying not to use timeouts) but that we will handle on our side.

Great work! :beers:

nojvek commented:
Done. 

Just curious: Any chance someone can give me a brief overview of how vscode uses es6 decorators to make dependency injection work?

```
		@IDebugService private debugService: IDebugService,
		@IContextMenuService private contextMenuService: IContextMenuService,
		@IInstantiationService private instantiationService: IInstantiationService,
		@IContextKeyService contextKeyService: IContextKeyService,
		@ICommandService private commandService: ICommandService,
		@ICodeEditorService private codeEditorService: ICodeEditorService,
		@ITelemetryService private telemetryService: ITelemetryService
```

Am I correct to assume that the services are all singletons? i.e each of them is only instantiated once and then stored in a service registry somewhere ?

```
@editorContribution
export class DebugEditorContribution implements IDebugEditorContribution {
```

This is like adding the class instance to another registry as editor extensions/contributors ? In this sense if I create another class

```
@editorContribution
export class SyncEditorContribution implements IDebugEditorContribution {
```

vscode will instantiate my contribution, add to contributions registry and call its handler functions when editor change happens right? So with `@editorContribution` its possible to build functionality in vscode that syncs the editor with a peer or web browser for a collaborative interview?

isidorn commented:
@nojvek thanks! I will merge it in today.
As for the code injection, yes every service is instantiated exactly once and more details can be found [here](https://github.com/Microsoft/vscode/wiki/Code-Organization#dependency-injection)

isidorn commented:
@nojvek added setting, started using set & map (since it is now ok to do this - we no longer support IE10) and did some polish on top. I also plan to look into restructuring the debugInlineValues.ts so it is a class and not a bunch of helper methods. I am also considering not showing inline values for lines after the current line that is focussed.
6091d5d
24c9eba
db334b2

nojvek commented:
Looks good. Thank you.

In the January release, This will be part of editor config right?

Can't wait to see this to completion.


On Fri, Jan 13, 2017 at 9:44 AM Isidor Nikolic <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> added setting, started using set &
> map (since it is now ok to do this - we no longer support IE10) and did
> some polish on top. I also plan to look into restructuring the
> debugInlineValues.ts so it is a class and not a bunch of helper methods. I
> am also considering not showing inline values for lines after the current
> line that is focussed.
>
>
> 6091d5d
> <https://github.com/Microsoft/vscode/commit/6091d5d77efda39c81c36ce2a0b753119ac1deed>
>
>
> 24c9eba
> <https://github.com/Microsoft/vscode/commit/24c9ebabefafd6ad518b7307c06d25b197fce25a>
>
>
> db334b2
> <https://github.com/Microsoft/vscode/commit/db334b22a6a4cbddd0b271e83aab98e2855a9d56>
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/vscode/pull/16129#issuecomment-272500268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-JVNbws0TEih9igNB44z872NZeZtT0ks5rR7gMgaJpZM4K9bza>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

isidorn commented:
This can be set in settings > "debug.inlineValues": true

luminaxster commented:
Wow!! Awesome work!
We are working on researching on and developing programming tools. What engine are you using to get the program state? Is it from scratch or using a VS API? Do you have some feedback(than can be shared) from the users that have used the feature? I highly recommend this [reading](http://www.cs.cmu.edu/~NatProg/papers/HCI%20in%20Tools%20Lifecycle%20-%20IEEE%20Comp%20PrePrint.pdf), if you haven't already, it goes "inline" with this feature.

weinand commented:
@luminaxster since this feature made its first appearance in VS Code 1.9 (3 days ago) we do not yet have statistics. In addition this feature is not enabled by default because we are still working on it, so we do not expect widespread acceptance in this release.
